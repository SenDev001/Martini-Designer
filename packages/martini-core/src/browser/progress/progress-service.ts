import { MessageService, Progress as TheiaProgress, ProgressMessage } from "@theia/core";
import { inject, injectable } from "inversify";

export const ProgressService = Symbol("ProgressService");

export interface ProgressService {
    showProgress<T>(message: ProgressMessage | string, task: (progress: Progress) => Promise<T>): Promise<T>;
}

export interface Progress extends TheiaProgress {
    isCancelled(): boolean;
}

@injectable()
export class DefaultProgressService implements ProgressService {
    @inject(MessageService) messageService: MessageService;

    async showProgress<T>(message: ProgressMessage | string, task: (progress: Progress) => Promise<T>): Promise<T> {
        let cancelled = false;
        const theiaProgress = await this.messageService.showProgress(
            typeof message === "string" ? { text: message } : message,
            () => (cancelled = true)
        );
        const progress = {
            ...theiaProgress,
            isCancelled() {
                return cancelled;
            }
        };
        try {
            return await task(progress);
        } finally {
            progress.cancel();
        }
    }
}
