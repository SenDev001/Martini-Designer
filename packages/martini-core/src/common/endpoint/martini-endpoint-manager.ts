import { JsonRpcServer } from "@theia/core";
import { MartiniEvent } from "../event/martini-event";

export const martiniEndpointManagerPath = "/services/martini/endpoints";

export enum EndpointType {
    DIR_WATCHER = "dir-watcher",
    EMAIL = "email",
    FTP_CLIENT = "ftp-client",
    FTP_SERVER = "ftp-server",
    HTTP_FILTER = "http-filter",
    JABBER = "jabber",
    JMS_LISTENER = "jms-listener",
    KAFKA_LISTENER = "kafka-listener",
    MONGODB_LISTENER = "mongodb-listener",
    URL_ALIAS = "rest-alias",
    RSS = "rss",
    SCHEDULER = "scheduler",
    TRACKER_RESUBMIT = "tracker-resubmit",
    REDIS_LISTENER = "redis-listener"
}

export interface MartiniEndpoint {
    type: EndpointType;
    name: string;
    service: string;
    packageName: string;
    enabled: boolean;
    modifiable: boolean;
    documentType: string;
    track: boolean;
    replicated: boolean;
    status: "STARTED" | "STOPPED";
}

export namespace MartiniEndpoint {
    export function is(object: any): object is MartiniEndpoint {
        return (
            !!object &&
            typeof object === "object" &&
            "type" in object &&
            "name" in object &&
            "service" in object &&
            "packageName" in object &&
            "enabled" in object &&
            "modifiable" in object &&
            "status" in object
        );
    }
    export function getType(type: string): EndpointType | undefined {
        const typeKey: string | undefined = Object.keys(EndpointType).find(
            (key: string) => EndpointType[key as keyof typeof EndpointType] === type
        );
        return typeKey ? EndpointType[typeKey as keyof typeof EndpointType] : undefined;
    }
}

export const MartiniEndpointManager = Symbol("MartiniEndpointManager");

export interface MartiniEndpointManager extends JsonRpcServer<MartiniEndpointManagerClient> {
    getAll(packageName: string): Promise<MartiniEndpoint[]>;

    get(packageName: string, endpointName: string): Promise<MartiniEndpoint | undefined>;

    start(packageName: string, endpointName: string): Promise<void>;

    stop(packageName: string, endpointName: string): Promise<void>;

    delete(packageName: string, endpointName: string): Promise<void>;

    save(packageName: string, endpoint: MartiniEndpoint): Promise<void>;

    test(packageName: string, endpoint: MartiniEndpoint): Promise<void>;
}

export interface EndpointEvent extends MartiniEvent {
    name: string;
    packageName: string;
    event: "STARTED" | "STOPPED" | "DELETED" | "SAVED";
}

export namespace EndpointEvent {
    export function is(event: MartiniEvent): event is EndpointEvent {
        return event.type === "ENDPOINT";
    }
}

export interface MartiniEndpointManagerClient {
    onEvent(event: EndpointEvent): void;
}

export const getDisplayName = (type: EndpointType): string => {
    switch (type) {
        case EndpointType.DIR_WATCHER:
            return "Directory Watcher";
        case EndpointType.EMAIL:
            return "Email";
        case EndpointType.KAFKA_LISTENER:
            return "Kafka Listener";
        case EndpointType.SCHEDULER:
            return "Scheduler";
        case EndpointType.TRACKER_RESUBMIT:
            return "Tracker Resubmit";
        case EndpointType.JABBER:
            return "Jabber";
        case EndpointType.FTP_CLIENT:
            return "FTP Client";
        case EndpointType.FTP_SERVER:
            return "FTP Server";
        case EndpointType.JMS_LISTENER:
            return "JMS Listener";
        case EndpointType.RSS:
            return "RSS";
        case EndpointType.URL_ALIAS:
            return "URL Alias";
        case EndpointType.MONGODB_LISTENER:
            return "MongoDB Listener";
        case EndpointType.REDIS_LISTENER:
            return "Redis Pub/Sub Listener";
        case EndpointType.HTTP_FILTER:
            return "HTTP Filter";
        default:
            return "Unknown";
    }
};

export interface RssEndpoint extends MartiniEndpoint {
    schedule: string;
    onlyNew: boolean;
    rssUrl: string;
}

export interface EmailSettings {
    host: string;
    port: number;
    username: string;
    password: string;
    ssl: boolean;
}

export interface ReplyEmailSettings extends EmailSettings {
    from: string;
}

export interface EmailEndpoint extends MartiniEndpoint, EmailSettings {
    schedule: string;
    replicated: boolean;
    track : boolean;
    protocol: "imap" | "pop3";
    deleteOnReceive: boolean;
    sendOutputAsReply: boolean;
    sendReplyOnError: boolean;
    replyEmailSettings: ReplyEmailSettings;
}
