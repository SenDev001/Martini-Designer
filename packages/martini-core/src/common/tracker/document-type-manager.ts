export interface DocumentType {
    id: string;
    name: string;
}

export const documentTypeManagerPath = "/services/martini/tracker/document-types";

export const DocumentTypeManager = Symbol("DocumentTypeManager");

export interface DocumentTypeManager {
    getAll(): Promise<DocumentType[]>;
    save(type: DocumentType): Promise<void>;
    delete(id: string): Promise<void>;
}
