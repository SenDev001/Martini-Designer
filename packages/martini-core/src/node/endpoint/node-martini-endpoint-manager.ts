import { inject, injectable, postConstruct } from "inversify";
import {
    EndpointEvent,
    MartiniEndpoint,
    MartiniEndpointManager,
    MartiniEndpointManagerClient
} from "../../common/endpoint/martini-endpoint-manager";
import { MartiniEventDispatcher } from "../event/martini-event-manager";
import { AxiosInstanceFactory } from "../http/axios-instance-factory";

@injectable()
export class MartiniEndpointManagerNode implements MartiniEndpointManager {
    @inject(AxiosInstanceFactory)
    private readonly axiosFactory: AxiosInstanceFactory;
    @inject(MartiniEventDispatcher)
    private readonly eventDispatcher: MartiniEventDispatcher;

    private client: MartiniEndpointManagerClient | undefined;
    private readonly excludedPayloadProperties = [
        "type",
        "name",
        "service",
        "packageName",
        "enabled",
        "modifiable",
        "status"
    ];

    @postConstruct()
    init() {
        this.eventDispatcher.onEvent(event => {
            if (EndpointEvent.is(event)) this.client?.onEvent(event);
        });
    }

    async getAll(packageName: string): Promise<MartiniEndpoint[]> {
        const response = await (await this.axiosFactory.make()).get(`/esbapi/packages/${packageName}/endpoints`, {
            params: { size: 1000 }
        });

        if (response.status !== 200) return [];

        return response.data.items.map((item: any) => ({
            type: item.type,
            name: item.name,
            service: item.service,
            packageName: item.packageName,
            enabled: item.enabled,
            modifiable: item.modifiable,
            status: item.status
        }));
    }

    async get(packageName: string, endpointName: string): Promise<MartiniEndpoint | undefined> {
        const response = await (await this.axiosFactory.make()).get(
            `/esbapi/packages/${packageName}/endpoints/${endpointName}`
        );

        if (response.status !== 200) return undefined;

        const data = response.data;

        return {
            type: data.type,
            name: data.name,
            service: data.service,
            packageName: data.packageName,
            enabled: data.enabled,
            modifiable: data.modifiable,
            status: data.status,
            ...data.properties
        };
    }

    async start(packageName: string, endpointName: string): Promise<void> {
        await (await this.axiosFactory.make()).put(`/esbapi/packages/${packageName}/endpoints/${endpointName}/start`);
    }

    async stop(packageName: string, endpointName: string): Promise<void> {
        await (await this.axiosFactory.make()).put(`/esbapi/packages/${packageName}/endpoints/${endpointName}/stop`);
    }

    async delete(packageName: string, endpointName: string): Promise<void> {
        const endpoint = await this.get(packageName, endpointName);

        if (endpoint && endpoint.status !== "STOPPED") await this.stop(packageName, endpointName);

        await (await this.axiosFactory.make()).delete(`/esbapi/packages/${packageName}/endpoints/${endpointName}`);
    }

    async save(packageName: string, endpoint: MartiniEndpoint): Promise<void> {
        await (await this.axiosFactory.make()).post(
            `/esbapi/packages/${packageName}/endpoints`,
            this.toSavePayload(endpoint)
        );
    }

    async test(packageName: string, endpoint: MartiniEndpoint): Promise<void> {
        await (await this.axiosFactory.make()).put(
            `/esbapi/packages/${packageName}/endpoints/validate`,
            this.toTestPayload(endpoint)
        );
    }

    private toSavePayload(endpoint: MartiniEndpoint) {
        return {
            type: endpoint.type,
            name: endpoint.name,
            endpointType: endpoint.name,
            endpointName: endpoint.name,
            service: endpoint.service,
            packageName: endpoint.packageName,
            enabled: endpoint.enabled,
            modifiable: endpoint.modifiable,
            properties: this.getProperties(endpoint)
        };
    }

    private toTestPayload(endpoint: MartiniEndpoint) {
        return {
            endpointName: endpoint.name,
            endpointType: endpoint.type,
            originalEndpointName: endpoint.name,
            enabled: endpoint.enabled,
            packageName: endpoint.packageName,
            serviceName: endpoint.service,
            properties: this.getProperties(endpoint)
        };
    }

    private getProperties(endpoint: MartiniEndpoint) {
        return Object.entries(endpoint)
            .filter(([key, _]) => !this.excludedPayloadProperties.includes(key))
            .reduce((properties, [key, value]) => {
                properties[key] = value;
                return properties;
            }, {} as { [key: string]: any });
    }

    dispose(): void {}

    setClient(client: MartiniEndpointManagerClient | undefined): void {
        this.client = client;
    }
}
