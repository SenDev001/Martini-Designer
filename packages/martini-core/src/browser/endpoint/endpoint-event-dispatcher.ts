import {Emitter, Event} from "@theia/core";
import {injectable} from "inversify";
import {EndpointEvent, MartiniEndpointManagerClient} from "../../common/endpoint/martini-endpoint-manager";

export const EndpointEventDispatcher = Symbol("EndpointEventDispatcher");

export interface EndpointEventDispatcher {
    onEndpointEvent: Event<EndpointEvent>;
}

@injectable()
export class EndpointEventDispatcherClient
    implements MartiniEndpointManagerClient, EndpointEventDispatcher {
    private readonly onEndpointEventEmitter = new Emitter<EndpointEvent>();
    readonly onEndpointEvent = this.onEndpointEventEmitter.event;

    onEvent(event: EndpointEvent): void {
        this.onEndpointEventEmitter.fire(event);
    }
}
