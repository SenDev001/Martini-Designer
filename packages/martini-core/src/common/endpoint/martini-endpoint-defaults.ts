import { EndpointType, MartiniEndpoint, RssEndpoint } from "./martini-endpoint-manager";

export const createDefaultEndpoint = (packageName: string, name: string, type: EndpointType): MartiniEndpoint => {
    switch (type) {
        case EndpointType.RSS:
            return createDefaultRssEndpoint(packageName, name);
        default:
            return {
                name,
                type,
                packageName,
                status: "STOPPED",
                enabled: true
            } as MartiniEndpoint;
    }
};

const createDefaultRssEndpoint = (packageName: string, name: string): RssEndpoint => ({
    name,
    type: EndpointType.RSS,
    packageName,
    service: "",
    modifiable: true,
    replicated: true,
    status: "STOPPED",
    documentType: "RSS",
    enabled: true,
    track: false,
    schedule: "repeating:1000",
    rssUrl: "",
    onlyNew: false
});
