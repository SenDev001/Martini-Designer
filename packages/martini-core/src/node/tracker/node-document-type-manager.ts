import { DocumentTypeManager, DocumentType } from "../../common/tracker/document-type-manager";
import { injectable, inject } from "inversify";
import { AxiosInstanceFactory } from "../http/axios-instance-factory";

@injectable()
export class DocumentTypeManagerNode implements DocumentTypeManager {
    @inject(AxiosInstanceFactory)
    private readonly axiosFactory: AxiosInstanceFactory;

    async getAll(): Promise<DocumentType[]> {
        const response = await (await this.axiosFactory.make()).get("/esbapi/tracker/document-types", {
            params: {
                page: 0,
                size: 1000
            }
        });

        return response.data.items as DocumentType[];
    }

    async save(type: DocumentType): Promise<void> {
        await (await this.axiosFactory.make()).post("/esbapi/tracker/document-types", type);
    }

    async delete(id: string): Promise<void> {
        await (await this.axiosFactory.make()).delete(`/esbapi/tracker/document-types/${id}`);
    }
}
