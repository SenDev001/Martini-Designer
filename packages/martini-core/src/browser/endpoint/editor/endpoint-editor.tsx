import { Disposable, Emitter, Event, MessageService } from "@theia/core";
import { Message, ReactWidget, Saveable, Widget, WidgetOpenerOptions, WidgetOpenHandler } from "@theia/core/lib/browser";
import URI from "@theia/core/lib/common/uri";
import { FormikErrors } from "formik";
import { inject, injectable, postConstruct } from "inversify";
import { cloneDeep, isEqualWith } from "lodash";
import messages from "martini-messages/lib/messages";
import * as React from "react";
import styled from "styled-components";
import { createDefaultEndpoint } from "../../../common/endpoint/martini-endpoint-defaults";
import { EndpointEvent, EndpointType, MartiniEndpoint, MartiniEndpointManager, RssEndpoint } from "../../../common/endpoint/martini-endpoint-manager";
import { DocumentTypeManager } from "../../../common/tracker/document-type-manager";
import { Loader } from "../../components/loader";
import { ConfirmDialog, createListMessage } from "../../dialogs/dialogs";
import { EditorFlag } from "../../editor/editor-flag";
import { convertFormikErrors, errorsToString, FormError } from "../../form/form";
import { FormSearchBox } from "../../form/form-search-box";
import { NavigatorOpenHandler } from "../../navigator/martini-navigator-view-widget";
import { ProgressService } from "../../progress/progress-service";
import { EndpointEventDispatcher } from "../endpoint-event-dispatcher";
import { EndpointEditorToolbar } from "./endpoint-editor-toolbar";
import { RssEndpointForm } from "./rss/rss-endpoint-form";

export const EndpointEditorOptions = Symbol("EndpointEditorOptions");
export interface EndpointEditorOptions {
    uri: string;
    endpointType: EndpointType;
    packageName: string;
    name: string;
    isNew: boolean;
}

const EndpointEditorStyles = styled.div`
    display: grid;
    grid-template-rows: max-content 1fr;
    height: 100%;

    .content {
        padding: var(--theia-ui-padding);
        overflow-y: auto;
    }
`;

@injectable()
export class EndpointEditor extends ReactWidget implements Saveable {
    static readonly FACTORY_ID = "endpoint-editor";
    static readonly URI_SCHEME = "endpoint";

    @inject(EndpointEditorOptions)
    private readonly options: EndpointEditorOptions;
    @inject(MartiniEndpointManager)
    private readonly endpointManager: MartiniEndpointManager;
    @inject(EndpointEventDispatcher)
    private readonly endpointEventDispatcher: EndpointEventDispatcher;
    @inject(ProgressService)
    private readonly progressService: ProgressService;
    @inject(MessageService)
    private readonly messageService: MessageService;
    @inject(DocumentTypeManager)
    private readonly documentTypeManager: DocumentTypeManager;

    private currentEndpoint: MartiniEndpoint;
    private initialEndpoint: MartiniEndpoint;

    private _dirty: boolean = false;
    readonly autoSave: "off";
    private readonly onDirtyChangedEmitter = new Emitter<void>();
    readonly onDirtyChanged: Event<void> = this.onDirtyChangedEmitter.event;
    private searchBox: FormSearchBox;
    private errors: FormError[] = [];
    private reset = false;

    @postConstruct()
    async init() {
        this.id = this.options.uri.toString();
        this.title.label = this.options.name;
        this.updateTitle();
        this.title.closable = true;
        this.node.tabIndex = -1;
        this.scrollOptions = undefined;
        this.searchBox = new FormSearchBox(this.node);
        EditorFlag.flag(this);

        if (!this.options.isNew) {
            this.update();
            this.initialEndpoint = await this.endpointManager.get(this.options.packageName, this.options.name) as MartiniEndpoint;
        }
        else {
            this.dirty = true;
            this.initialEndpoint = createDefaultEndpoint(this.options.packageName, this.options.name, this.options.endpointType);
        }

        this.currentEndpoint = cloneDeep(this.initialEndpoint);

        this.toDispose.push(this.endpointEventDispatcher.onEndpointEvent(this.handleEndpointEvent));
        this.update();
    }

    get dirty() {
        return this._dirty;
    }

    set dirty(dirty: boolean) {
        if (this._dirty !== dirty) {
            this._dirty = dirty;
            this.onDirtyChangedEmitter.fire();
        }
    }

    async save() {
        if (this.errors.length === 0) {
            await this.progressService.showProgress(messages.endpoint_editor_saving, async () => {
                try {
                    let start = false;
                    if (this.currentEndpoint.status === "STARTED") {
                        start = true;
                        await this.endpointManager.stop(this.currentEndpoint.packageName, this.currentEndpoint.name);
                    }
                    await this.endpointManager.save(this.currentEndpoint.packageName, this.currentEndpoint);
                    this.initialEndpoint = this.currentEndpoint;
                    if (this.options.isNew) {
                        this.options.isNew = false;
                        this.update();
                    }
                    this.dirty = false;
                    if (start)
                        await this.endpointManager.start(this.currentEndpoint.packageName, this.currentEndpoint.name);
                } catch (error) {
                    this.messageService.error(error.toString());
                }
            });
        } else {
            const dlg = new ConfirmDialog({
                title: messages.unable_to_save_title,
                showCancel: false,
                msg: createListMessage(
                    messages.unable_to_save,
                    this.errors.map(({ label, message }) => `${label}: ${message}`)
                )
            });
            await dlg.open();
        }
    }

    public render(): React.ReactNode {
        if (this.currentEndpoint) {
            return <EndpointEditorStyles tabIndex={-1}>
                <EndpointEditorToolbar
                    onStart={this.options.isNew ? undefined : this.handleStart}
                    onStop={this.options.isNew ? undefined : this.handleStop}
                    onTest={this.handleTestConfiguration}
                    onRevert={this.handleRevert}
                    onSearch={this.handleSearch}
                    endpointName={this.options.name}
                    endpointType={this.options.endpointType}
                    status={this.currentEndpoint.status}
                    errors={this.errors}
                />
                <div className="content">
                    {this.renderForm()}
                </div>
            </EndpointEditorStyles>;
        }
        else
            return <Loader message={messages.loading} />;
    }

    showSearchBox() {
        this.handleSearch();
    }

    onAfterAttach(msg: Message) {
        super.onAfterAttach(msg);

        if (this.searchBox.isAttached)
            Widget.detach(this.searchBox);
        Widget.attach(this.searchBox, this.node);
        this.node.addEventListener('keydown', e => {
            if (e.metaKey && e.key === 'f')
                this.handleSearch();
        });
        this.toDisposeOnDetach.push(Disposable.create(() => Widget.detach(this.searchBox)));
    }

    private handleChange(endpoint?: MartiniEndpoint) {
        if (endpoint)
            this.currentEndpoint = endpoint;
        this.dirty = this.isDirty();
    }

    // @ts-ignore
    private handleErrors(errors: FormikErrors<MartiniEndpoint>) {
        this.errors = convertFormikErrors(this.node, errors);
        this.updateTitle();
        this.update();
    }

    private handleTestConfiguration = async () => {
        this.progressService.showProgress(messages.endpoint_editor_testing, async () => {
            try {
                await this.endpointManager.test(this.currentEndpoint.packageName, this.currentEndpoint);
                this.messageService.info(messages.endpoint_editor_testing_success, {
                    timeout: 2000
                });
            } catch (error) {
                console.error(error);
                this.messageService.error(error.toString());
            }
        });
    };

    private handleStart = async () => {
        if (this.currentEndpoint.status === "STARTED")
            return;
        this.progressService.showProgress(messages.endpoint_editor_starting, async () => {
            try {
                await this.endpointManager.start(this.options.packageName, this.options.name);
            } catch (error) {
                console.error(error);
                this.messageService.error(error.toString());
            }
        });
    };

    private handleStop = async () => {
        if (this.currentEndpoint.status === "STOPPED")
            return;
        this.progressService.showProgress(messages.endpoint_editor_stopping, async () => {
            try {
                await this.endpointManager.stop(this.options.packageName, this.options.name);
            } catch (error) {
                console.error(error);
                this.messageService.error(error.toString());
            }
        });
    };

    private handleRevert = async () => {
        const dlg = new ConfirmDialog({
            title: messages.revert_all_title,
            msg: messages.revert_all_question,
            ok: messages.revert
        });

        const revert = await dlg.open();
        if (revert) {
            this.currentEndpoint = this.initialEndpoint;
            this.reset = true;
            this.handleChange();
            this.update();
        }
    };

    private handleSearch = () => {
        this.searchBox.show();
        this.searchBox.activate();
    };

    private handleEndpointEvent = async (event: EndpointEvent) => {
        if (event.event === "DELETED") {
            this.close();
            return;
        }

        if (event.event === "STARTED" || event.event === "STOPPED") {
            this.currentEndpoint.status = event.event;
            this.update();
        } else if (event.event === "SAVED") {
            const updatedEndpoint = await this.endpointManager.get(this.options.packageName, this.options.name);
            if (!this.dirty) {
                this.currentEndpoint = updatedEndpoint!;
                this.reset = true;
                this.update();
            }
            this.initialEndpoint = updatedEndpoint!;
        }
    };

    private isDirty(): boolean {
        return this.options.isNew || !isEqualWith(this.initialEndpoint, this.currentEndpoint, (_value, _other, indexOrKey) => {
            if (indexOrKey === "status")
                return true;
        });
    }

    private updateTitle() {
        this.title.label = this.options.name;

        if (this.errors.length === 0) {
            this.title.caption = `Endpoint - ${this.options.name}`;
            this.title.iconClass = `martini-tab-icon martini-${this.options.endpointType}-endpoint-icon`;
        }
        else {
            this.title.iconClass = "martini-tab-icon martini-error-icon";
            this.title.caption = errorsToString(this.errors);
        }
    }

    private renderForm(): React.ReactNode {
        const _reset = this.reset;
        this.reset = false;
        switch (this.options.endpointType) {
            case EndpointType.RSS:
                return <RssEndpointForm
                    endpoint={this.currentEndpoint as RssEndpoint}
                    documentTypeProvider={() => this.documentTypeManager.getAll()}
                    onChange={endpoint => this.handleChange(endpoint)}
                    onValidate={errors => this.handleErrors(errors)}
                    reset={_reset}
                />;
            default:
                return <></>;
        }
    }
}

@injectable()
export class EndpointEditorNavigatorOpenHandler implements NavigatorOpenHandler {

    getUri(element: any): URI {
        return new URI(`${EndpointEditor.URI_SCHEME}://${encodeURI(element.packageName)}/${encodeURI(element.name)}?type=${element.type}`);
    }

    canHandle(element: any): boolean {
        return MartiniEndpoint.is(element);
    }

}

@injectable()
export class EndpointEditorOpenHandler extends WidgetOpenHandler<EndpointEditor> {
    readonly id = EndpointEditor.FACTORY_ID;

    canHandle(uri: URI, _?: WidgetOpenerOptions): number {
        return uri.scheme === EndpointEditor.URI_SCHEME ? 200 : -1;
    }

    protected createWidgetOptions(uri: URI, _?: WidgetOpenerOptions): EndpointEditorOptions {
        const params: { [key: string]: string; } = {};
        uri.query.split("&").forEach(param => {
            const [key, value] = param.split("=");
            params[key] = value;
        });
        return {
            uri: uri.withoutFragment().toString(),
            endpointType: MartiniEndpoint.getType(params["type"])!,
            name: uri.path.base,
            packageName: decodeURI(uri.authority),
            isNew: params["new"] === "true",
        };
    }
}
