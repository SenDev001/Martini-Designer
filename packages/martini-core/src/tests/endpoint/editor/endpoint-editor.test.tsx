require("reflect-metadata");
import { MessageClient } from "@theia/core";
import { MessageService } from "@theia/core/lib/common/message-service";
import * as Enzyme from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import { Container } from "inversify";
import { noop } from "lodash";
import * as React from "react";
import * as Renderer from "react-test-renderer";
import { EndpointEditor, EndpointEditorOptions } from "../../../browser/endpoint/editor/endpoint-editor";
import { EndpointEventDispatcher, EndpointEventDispatcherClient } from "../../../browser/endpoint/endpoint-event-dispatcher";
import { ProgressService } from "../../../browser/progress/progress-service";
import { createDefaultEndpoint } from "../../../common/endpoint/martini-endpoint-defaults";
import { EndpointType, MartiniEndpointManager } from "../../../common/endpoint/martini-endpoint-manager";
import { MartiniEndpointManagerNode } from "../../../node/endpoint/node-martini-endpoint-manager";
import { DocumentTypeManager } from "../../../common/tracker/document-type-manager";
import { DocumentTypeManagerNode } from "../../../node/tracker/node-document-type-manager";

jest.mock("../../../node/endpoint/node-martini-endpoint-manager");
jest.mock("../../../node/tracker/node-document-type-manager");
jest.mock("@theia/core/src/common/message-service");

Enzyme.configure({ adapter: new Adapter() });

const container = new Container();
container.bind(MartiniEndpointManager).toConstantValue(new MartiniEndpointManagerNode());
container.bind(DocumentTypeManager).toConstantValue(new DocumentTypeManagerNode());
container.bind(ProgressService).toConstantValue({
    showProgress: (_, task: any) =>
        task({
            isCancelled: () => false,
            report: noop
        })
} as ProgressService);
container.bind(MessageService).toConstantValue(new MessageService({} as MessageClient));
container.bind(EndpointEventDispatcher).to(EndpointEventDispatcherClient);
container.bind(EndpointEditor).toSelf().inRequestScope();
const options: EndpointEditorOptions = {
    uri: "endpoint://test",
    isNew: false,
    endpointType: EndpointType.DIR_WATCHER,
    name: "test",
    packageName: "test"
};
container.bind(EndpointEditorOptions).toConstantValue(options);
const manager = container.get<MartiniEndpointManager>(MartiniEndpointManager);

test("Editor should be rendered with new endpoint", async () => {
    options.isNew = true;
    const editor = container.get(EndpointEditor);
    const node = Renderer.create(editor.render() as React.ReactElement);
    const tree = node.toJSON();
    expect(tree).toMatchSnapshot();
    expect(editor.dirty).toBe(true);
});

test("Editor should be rendered with existing endpoint", async () => {
    options.isNew = false;
    const editor = container.get(EndpointEditor);
    manager.get = jest.fn(async name => createDefaultEndpoint("test", name, EndpointType.DIR_WATCHER));
    await Renderer.act(async () => await editor.init());
    expect(manager.get).toBeCalledWith("test", "test");
    const node = Renderer.create(editor.render() as React.ReactElement);
    const tree = node.toJSON();
    expect(tree).toMatchSnapshot();
    expect(editor.dirty).toBe(false);
});

test("Saving should make the editor not dirty", async () => {
    options.isNew = true;
    const editor = container.get(EndpointEditor);
    manager.save = jest.fn(_ => Promise.resolve());
    expect(editor.dirty).toBe(true);
    Object.assign(editor, {
        errors: []
    });
    await editor.save();
    expect(manager.save).toBeCalledTimes(1);
    expect(editor.dirty).toBe(false);
});

test("Should restart the endpoint after saving if started", async () => {
    const editor = container.get(EndpointEditor);
    manager.get = jest.fn(async name => {
        const endpoint = createDefaultEndpoint("test", name, EndpointType.DIR_WATCHER);
        endpoint.status = "STARTED";
        return endpoint;
    });

    await editor.init();
    Object.assign(editor, {
        errors: []
    });
    manager.save = jest.fn(_ => Promise.resolve());
    manager.start = jest.fn(_ => Promise.resolve());
    manager.stop = jest.fn(_ => Promise.resolve());
    Object.assign(editor, {
        errors: []
    });
    await editor.save();
    expect(manager.save).toBeCalledTimes(1);
    expect(manager.stop).toBeCalledTimes(1);
    expect(manager.start).toBeCalledTimes(1);
    expect(editor.dirty).toBe(false);
});
