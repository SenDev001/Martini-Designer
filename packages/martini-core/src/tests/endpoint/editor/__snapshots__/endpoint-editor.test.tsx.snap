// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor should be rendered with existing endpoint 1`] = `
<div
  className="sc-AxgMl jA-DMuo"
  tabIndex={-1}
>
  <div
    className="sc-AxhCb ilrYiD"
  >
    <div
      className="martini-tree-icon martini-dir-watcher-endpoint-icon"
      style={
        Object {
          "marginRight": "4px",
          "position": "relative",
        }
      }
    >
      <div
        className="martini-stopped-overlay-icon"
        style={
          Object {
            "backgroundPosition": "center center",
            "backgroundRepeat": "no-repeat",
            "backgroundSize": "7px 8px",
            "height": 8,
            "position": "absolute",
            "right": "0",
            "top": "0",
            "width": 7,
          }
        }
      />
    </div>
    <div
      className="title-wrapper"
    >
      <div
        className="title"
      >
        Directory Watcher
         Connection - 
        test
      </div>
    </div>
    <div
      className="sc-AxiKw bbpvGZ"
      style={
        Object {
          "gridTemplateColumns": "repeat(5, max-content)",
          "height": undefined,
        }
      }
    >
      <button
        className="theia-button"
        disabled={false}
        onClick={[Function]}
      >
        Test Configuration
      </button>
      <button
        className="theia-button"
        onClick={[Function]}
      >
        Revert
      </button>
      <div
        className="item item-enabled"
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseOut={[Function]}
        onMouseUp={[Function]}
        title="Find (Ctrl+F)"
      >
        <div
          className="martini-icon martini-search-icon"
          style={
            Object {
              "width": "16px",
            }
          }
        />
      </div>
      <div
        className="item item-enabled"
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseOut={[Function]}
        onMouseUp={[Function]}
      >
        <div
          className="martini-icon martini-start-icon"
          style={
            Object {
              "width": "16px",
            }
          }
        />
      </div>
      <div
        className="item"
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseOut={[Function]}
        onMouseUp={[Function]}
      >
        <div
          className="martini-icon martini-stop-icon disabled-icon"
          style={
            Object {
              "width": "16px",
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="content"
  />
</div>
`;

exports[`Editor should be rendered with new endpoint 1`] = `
<div
  className="sc-AxgMl jA-DMuo"
  tabIndex={-1}
>
  <div
    className="sc-AxhCb ilrYiD"
  >
    <div
      className="martini-tree-icon martini-dir-watcher-endpoint-icon"
      style={
        Object {
          "marginRight": "4px",
          "position": "relative",
        }
      }
    >
      <div
        className="martini-stopped-overlay-icon"
        style={
          Object {
            "backgroundPosition": "center center",
            "backgroundRepeat": "no-repeat",
            "backgroundSize": "7px 8px",
            "height": 8,
            "position": "absolute",
            "right": "0",
            "top": "0",
            "width": 7,
          }
        }
      />
    </div>
    <div
      className="title-wrapper"
    >
      <div
        className="title"
      >
        Directory Watcher
         Connection - 
        test
      </div>
    </div>
    <div
      className="sc-AxiKw bbpvGZ"
      style={
        Object {
          "gridTemplateColumns": "repeat(5, max-content)",
          "height": undefined,
        }
      }
    >
      <button
        className="theia-button"
        disabled={false}
        onClick={[Function]}
      >
        Test Configuration
      </button>
      <button
        className="theia-button"
        onClick={[Function]}
      >
        Revert
      </button>
      <div
        className="item item-enabled"
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseOut={[Function]}
        onMouseUp={[Function]}
        title="Find (Ctrl+F)"
      >
        <div
          className="martini-icon martini-search-icon"
          style={
            Object {
              "width": "16px",
            }
          }
        />
      </div>
      <div
        className="item"
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseOut={[Function]}
        onMouseUp={[Function]}
      >
        <div
          className="martini-icon martini-start-icon disabled-icon"
          style={
            Object {
              "width": "16px",
            }
          }
        />
      </div>
      <div
        className="item"
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseOut={[Function]}
        onMouseUp={[Function]}
      >
        <div
          className="martini-icon martini-stop-icon disabled-icon"
          style={
            Object {
              "width": "16px",
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="content"
  />
</div>
`;
